Groovy 5 compatibility

matrei opened the pull request:
This is the start of checking compatibility with Groovy 5 and eventually adding a CI workflow for Groovy 5.0.0-SNAPSHOT.

Superseded by https://github.com/apache/grails-core/pull/15183

matrei committed:
build: Groovy 5.0.0-SNAPSHOT and Spock 2.4-M6-groovy-4.0

matrei committed:
build: add `-Dspock.iKnowWhatImDoing.disableGroovyVersionCheck=true`

matrei committed:
fix: Groovy 5 compatibility

matrei committed:
test: Add `@PendingFeatureIf` for currently incompatible tests

matrei committed:
Merge branch '7.0.x' into groovy-5

# Conflicts:
#	dependencies.gradle
#	gradle/functional-test-config.gradle
#	grails-gradle/plugins/src/e2eTest/groovy/org/grails/gradle/test/GrailsPublishPluginSpec.groovy
#	grails-test-examples/async-events-pubsub-demo/src/integration-test/groovy/pubsub/demo/TaskControllerSpec.groovy

matrei committed:
fix: replace deprecated `capitalize` method

Groovy's `MetaClassHelper.capitalize` method
has been deprecated for removal since Groovy 3.

jdaugherty commented:
I have been experimenting with a local snapshot of Spock fixed with Eric's suggestions here: https://github.com/spockframework/spock/discussions/2160#discussioncomment-14186360

One of the issues I found was: 

    1: Task failed with an exception.
    -----------
    * What went wrong:
    Execution failed for task ':grails-datastore-core:compileGroovy'.
    > Unrecoverable compilation error: startup failed:
      General error during conversion: The interface Collection cannot be implemented more than once with different arguments: java.util.Collection<java.lang.Object> and java.util.Collection
      . At [30:1] 
      
      org.codehaus.groovy.syntax.RuntimeParserException: The interface Collection cannot be implemented more than once with different arguments: java.util.Collection<java.lang.Object> and java.util.Collection
      . At [30:1] 
            at org.codehaus.groovy.classgen.Verifier.checkForDuplicateInterfaces(Verifier.java:409)
            at org.codehaus.groovy.classgen.Verifier.visitClass(Verifier.java:235)
            at org.codehaus.groovy.tools.javac.JavaStubGenerator$1.visitClass(JavaStubGenerator.java:236)
            at org.codehaus.groovy.tools.javac.JavaStubGenerator.printClassContents(JavaStubGenerator.java:331)
            at org.codehaus.groovy.tools.javac.JavaStubGenerator.generateStubContent(JavaStubGenerator.java:204)
            at org.codehaus.groovy.tools.javac.JavaStubGenerator.generateFileStub(JavaStubGenerator.java:181)
            at org.codehaus.groovy.tools.javac.JavaStubGenerator.generateClass(JavaStubGenerator.java:161)
            at org.codehaus.groovy.tools.javac.JavaAwareCompilationUnit.lambda$new$2(JavaAwareCompilationUnit.java:111)
            at org.codehaus.groovy.control.CompilationUnit$IPrimaryClassNodeOperation.doPhaseOperation(CompilationUnit.java:920)
            at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:666)
            at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:640)
            at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:621)
            at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:285)
            at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:67)
            at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:115)
            at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:99)
            at org.gradle.api.internal.tasks.compile.daemon.AbstractIsolatedCompilerWorkerExecutor$CompilerWorkAction.execute(AbstractIsolatedCompilerWorkerExecutor.java:78)
            at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
            at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:54)
            at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:48)
            at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
            at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:48)
            at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
            at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
            at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:108)
            at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:77)
            at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$1(WorkerAction.java:150)
            at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
            at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$2(WorkerAction.java:150)
            at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
            at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:142)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
            at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.base/java.lang.reflect.Method.invoke(Method.java:569)
            at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
            at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
            at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
            at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
            at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
            at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
            at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
            at java.base/java.lang.Thread.run(Thread.java:840)
      
      1 error

    ==============================================================================

matrei commented:
@jdaugherty Yes, I have created a ticket for that here: https://issues.apache.org/jira/browse/GROOVY-11736

matrei committed:
Merge branch '7.0.x' into groovy-5

# Conflicts:
#	dependencies.gradle

matrei committed:
fix: Groovy 5 compatibility

https://issues.apache.org/jira/browse/GROOVY-11736

matrei commented:
This is an issue with Groovy 5.0.0:
https://issues.apache.org/jira/projects/GROOVY/issues/GROOVY-11743

matrei committed:
test: add `@PendingFeatureIf` for failing tests

matrei committed:
test: remove `@PendingFeatureIf` for passing tests

These test now pass with Groovy 5.0.1

matrei committed:
fix: set required `value` on `@Generated`

This fixes compilation error:
No explicit/default value found for annotation attribute
'value' in @jakarta.annotation.Generated

matrei committed:
fix(deps): `groovy` v. `5.0.0-SNAPSHOT` -> `5.0.1-SNAPSHOT`

matrei committed:
Merge branch '7.0.x' into fork/matrei/groovy-5

# Conflicts:
#	grails-bootstrap/src/main/groovy/org/grails/config/NavigableMap.groovy
#	grails-core/src/main/groovy/org/grails/compiler/injection/AbstractGrailsArtefactTransformer.java
#	grails-core/src/main/groovy/org/grails/compiler/injection/GrailsASTUtils.java
#	grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/reflect/NameUtils.java

matrei committed:
chore: formatting

matrei committed:
test: add `@PendingFeatureIf` for failing tests

jdaugherty commented:
I merged 7.0.x into this branch, but there are 2 errors that stand out: 

1. Code Style - codenarc is failing to detect a property defined on the parent as an implementation of the abstract method from a child  (See CreatePluginCommand failure)

2. GormEntityDirtyCheckable uses Jakarta's `@Generated` and is saying the `value` attribute is required now.

jdaugherty committed:
fix: switch to groovy.transform.Generated

jdaugherty commented:
GormEntityDirtyCheckable is fixed by adopting the Generated annotation from Groovy - like we use everywhere else.

jdaugherty commented:
Running compileGroovy causes a failure similar to code narc's failure: 

	grails-core/grails-core/src/main/groovy/grails/boot/config/tools/ProfilingGrailsApplicationPostProcessor.groovy: 35: Can't have an abstract method in a non-abstract class. The class 'grails.boot.config.tools.ProfilingGrailsApplicationPostProcessor' must be declared abstract or the method 'void onApplicationEvent(E)' must be implemented.
	 @ line 35, column 1.
	   class ProfilingGrailsApplicationPostProcessor extends GrailsApplicationPostProcessor implements BeanPostProcessor {
	   ^

	1 error

But GrailsApplicationPostProcessor implements an onApplicationEvent() so I'm not sure why Groovy is failing to find the inherited method from the parent.  

I tried variations on this code here: https://github.com/jdaugherty/groovy5.0.1-inheritance-problem - eventually I even included the spring classes and it doesn't fail to compile.  There's some edge case I'm missing.

To test this: `./gradlew :grails-core:compileGroovy`

matrei committed:
fix: remove value from Groovy `@Generated`

matrei committed:
fix: switch order of checks in condition

See https://issues.apache.org/jira/browse/GROOVY-11754

matrei committed:
build: add Groovy version flag for Spock in forge

matrei committed:
fix(deps): `groovy` v. `5.0.1-SNAPSHOT` -> `5.0.2-SNAPSHOT`

matrei committed:
Merge branch '7.0.x' into groovy-5

jdaugherty commented:
@matrei looks like groovy 5 support has been added to the spock code line.

matrei commented:
Superseded by #15183
